File: linkedlist/doubly_linkedlist.py
class DNode:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None
        
class DoublyLinkedList:
    
    def __init__(self):
        self.head = None
        
    def insertAtHead(self, d):
        temp = DNode(d)
        if self.head == None:
            self.head = temp
            return
        temp.next = self.head
        self.head.prev = temp
        self.head = temp
        
    def insertAtTail(self, d):
        temp1 = DNode(d)
        if self.head == None:
            self.head = temp1
            return
        temp2 = self.head
        while temp2.next != None:
            temp2 = temp2.next
        temp2.next = temp1
        temp1.prev = temp2
        
    def printList(self):
        if self.head == None:
            print "List is empty!"
            return
        temp = self.head
        while temp != None:
            print str(temp.data) + "-->",
            temp = temp.next
        print
        
    def printReverse(self):
        if self.head == None:
            print "List is empty!"
            return
        temp = self.head
        while temp.next != None:
            temp = temp.next
        while temp != None:
            print str(temp.data) + "-->",
            temp = temp.prev
        print
        
if __name__ == '__main__':
    dlist = DoublyLinkedList()
    dlist.insertAtHead(1)
    dlist.insertAtTail(2)
    dlist.insertAtTail(3)
    dlist.insertAtTail(4)
    dlist.printList()
    dlist.printReverse()
    

File: JavaCodes/src/linkedlists/DoublyLinkedList.java
package linkedlists;

public class DoublyLinkedList {
	
	DNode head;
	
	public DoublyLinkedList(){
		head = null;
	}

	public void insertAtHead(Object d){
		DNode temp = new DNode(d);
		if (head == null){
			head = temp;
			return;
		}
		temp.next = head;
		head.prev = temp;
		head = temp;
	}
	
	public void insertAtTail(Object d){
		DNode temp1 = new DNode(d);
		if (head == null){
			head = temp1;
			return;
		}
		DNode temp2 = head;
		while(temp2.next != null){
			temp2 = temp2.next;
		}
		temp2.next = temp1;
		temp1.prev = temp2;
		
	}
	
	public void printList(){
		DNode temp = head;
		if(temp == null){
			System.out.println("List is empty!");
			return;
		}
		while(temp != null){
			System.out.print((Integer)temp.data + "->"); 
			temp = temp.next;
		}
		System.out.println();
	}
	
	public void reversePrint(){
		DNode temp = head;
		if(temp == null){
			System.out.println("List is empty!");
			return;
		}
		while(temp.next != null){
			temp = temp.next;
		}
		while(temp != null){
			System.out.print((Integer)temp.data + "->"); 
			temp = temp.prev;
		}
		System.out.println();
	}
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		DoublyLinkedList dl = new DoublyLinkedList();
		dl.printList();
		//dl.insertAtHead(1);
		//dl.insertAtHead(5);
		//dl.insertAtHead(7);
		//dl.printList();
		//dl.reversePrint();
		dl.insertAtTail(54);
		dl.insertAtTail(68);
		dl.printList();
	}

}
